### PR TITLE
x-vector: Add 'embed', 'concatWith', 'chunksOf' for streaming

### PR DESCRIPTION
**Pull request by jystic**

### CODE REVIEW

File: x-vector/src/X/Data/Vector/Stream/Group.hs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+concatWith :: Monad m => Stream m a -> (a -> Stream m b) -> Stream m b
+concatWith =
+  flip Stream.concatMap
+{-# INLINE [1] concatWith #-}
+
+chunksOf :: (PrimMonad b, MonadBase b m, Generic.Vector v a) => Int -> Stream m a -> Stream m (v a)
+chunksOf !n (Stream step t) =
+  if n <= 0 then
+    Savage.error "X.Data.Vector.Stream.Group.chunksOf: chunk size must be greater than zero"
```
Comment (jystic) [inline x-vector/src/X/Data/Vector/Stream/Group.hs]: Not sure what to do about this, I could alternatively yield chunks of 1, but perhaps that indicates a bug?

File: x-vector/src/X/Data/Vector/Stream/Group.hs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  else
+    embed $ do
+      let
+        loop = \case
+          Nothing' ->
+            pure Done
+
+          Just' (s0, i, g0) ->
+            if i == n then do
```
Comment (tranma) [inline x-vector/src/X/Data/Vector/Stream/Group.hs]: should this just be `i <= n`? (if I'm reading correctly)

File: x-vector/test/Test/X/Data/Vector/Stream/Group.hs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    let
+      yss0 =
+        Stream.chunksOf n $
+        Stream.fromList xs0
+
+    yss <- Stream.toList yss0
+
+    pure . counterexample ("chunks = " <> show yss) . conjoin $
+      [ concatMap Unboxed.toList yss === xs0 ] <>
+      fmap (\ys -> property $ Unboxed.length ys <= n && not (Unboxed.null ys)) yss
```
Comment (tranma) [inline x-vector/test/Test/X/Data/Vector/Stream/Group.hs]: a comment on each of these properties might make it easier to skim :)

File: x-vector/src/X/Data/Vector/Stream/Group.hs (hunk 1 above)
Comment (tranma) [inline x-vector/src/X/Data/Vector/Stream/Group.hs]: as long as it treats invalid n's as errors, I'm not fussed

File: x-vector/src/X/Data/Vector/Stream/Group.hs (hunk 2 above)
Comment (tranma) [inline x-vector/src/X/Data/Vector/Stream/Group.hs]: nvm, `i` increases up to `n` 👍

**Comment by tranma** [comment]
🍊

**Pull request merged**